add zmq_ppoll

egpbos opened the pull request:
zmq_ppoll mostly mimics zmq_poll behavior, except for the added feature of being able to specify a signal mask. Signals in this mask will be blocked during execution of zmq_ppoll. Switching of the process' active signal mask happens atomically with the actual poll call, so that no race conditions can occur. This behavior is useful when one wants to gracefully handle POSIX signals without race conditions. See e.g. the discussion below https://250bpm.com/blog:12/ for an explanation.

Also includes two new tests:
1. test_zmq_ppoll_fd does the same thing as test_zmq_poll_fd, demonstrating backwards compatibility with zmq_poll when used with a default signal mask.
2. test_zmq_ppoll_signals demonstrates the use of zmq_ppoll with a signal mask, blocking out SIGTERM everywhere except in zmq_ppoll, allowing to handle the signal in one place without having to worry about race conditions.

Fixes #4220.

Edit: note for reviewers: to prevent code duplication, I factored out some common parts of `zmq_poll` so that they could be reused in `zmq_ppoll`.

bluca commented on include/zmq.h:
```diff
@@ -513,6 +516,7 @@ typedef struct zmq_pollitem_t
 #define ZMQ_POLLITEMS_DFLT 16
 
 ZMQ_EXPORT int zmq_poll (zmq_pollitem_t *items_, int nitems_, long timeout_);
+ZMQ_EXPORT int zmq_ppoll (zmq_pollitem_t *items_, int nitems_, long timeout_, const sigset_t *sigmask_ = nullptr);
```
move to draft section, and add ifdefs as it will break windows builds

bluca commented on tests/test_zmq_ppoll_fd.cpp:
```diff
@@ -0,0 +1,90 @@
+/*
+    Copyright (c) 2016 Contributors as noted in the AUTHORS file
```
2021

bluca commented on tests/test_zmq_ppoll_signals.cpp:
```diff
@@ -0,0 +1,241 @@
+{
+    UNITY_BEGIN ();
+    RUN_TEST (test_ppoll_signals);
+//    RUN_TEST (test_poll_signals);
```
don't add commented out stuff

egpbos replied:
It was meant as an illustration for devs of why it is necessary. I will remove it.

bluca commented on src/polling_util.hpp:
```diff
@@ -108,8 +108,8 @@ typedef int timeout_t;
 compute_timeout (bool first_pass_, long timeout_, uint64_t now_, uint64_t end_);
-
-#elif defined ZMQ_POLL_BASED_ON_SELECT
+#endif
```
this does not look right

egpbos replied:
I think there are checks to exclude the possibility of ZMQ_POLL_BASED_ON_POLL and ZMQ_POLL_BASED_ON_SELECT being defined at the same time, so I think it makes no difference for those cases. I changed it like this, because ZMQ_POLL_BASED_ON_POLL can be true at the same time as ZMQ_HAVE_PPOLL, because ppoll is currently based purely on pselect, so it is independent of the poller chosen for existing poll functions.

egpbos replied:
I made sure that the `elif` behavior is still obeyed by replacing the following `if` clause with `#if (!defined ZMQ_POLL_BASED_ON_POLL && defined ZMQ_POLL_BASED_ON_SELECT) || defined ZMQ_HAVE_PPOLL`.

Note that the elif is not possible, because then the block is always skipped when `ZMQ_POLL_BASED_ON_POLL`, but the block is also necessary in case of `ZMQ_HAVE_PPOLL`.

bluca commented on src/zmq.cpp:
```diff
@@ -846,25 +851,10 @@ static int zmq_poller_poll (zmq_pollitem_t *items_, int nitems_, long timeout_)
 }
 #endif // ZMQ_HAVE_POLLER
 
-int zmq_poll (zmq_pollitem_t *items_, int nitems_, long timeout_)
```
This is a new function, why are you changing all the existing other implementations? Leave them as they are

egpbos replied:
Most of `zmq_poll` and `zmq_ppoll` are identical, the only big difference is the `pselect` call which needs an extra `sigmask` parameter. Factoring out the identical parts prevents code duplication. I can revert if you prefer, of course, but I had imagined it would create an unnecessary maintenance burden.

bluca replied:
avoiding to break the stable api is more important

egpbos replied:
Reverted `zmq_poll`.

bluca commented:
autoconf checks are missing

egpbos commented:
Thanks for the super quick review! Do you prefer rebase/fixup to keep commit history clean? Or should I add a new "fix review comments" commit?

bluca commented:
rebase

egpbos commented:
Can you give me a pointer on where I should add autoconf checks? I'm not so familiar with autoconf.

bluca commented:
grep for the same definitions and you'll see them in acinclude configure.ac et al

bluca commented on include/zmq.h:
```diff
@@ -763,6 +766,10 @@ ZMQ_EXPORT int zmq_socket_monitor_versioned (
   void *s_, const char *addr_, uint64_t events_, int event_version_, int type_);
 ZMQ_EXPORT int zmq_socket_monitor_pipes_stats (void *s);
 
+#ifdef ZMQ_HAVE_PPOLL
```
Don't ifdef in the public header, otherwise the ABI is different depending on the build. Instead return an error OPNOTSUPP or something like that from the implementation

egpbos commented:
Addressed some comments, still have to do the autoconf checks.

egpbos commented:
I added everything to the autoconf build system as well now. Compiles and passes tests on my system, but I'm sure there will be failures on others, as I have little experience with autoconf. Please review :)

bluca commented on include/zmq.h:
```diff
@@ -763,6 +766,8 @@ ZMQ_EXPORT int zmq_socket_monitor_versioned (
   void *s_, const char *addr_, uint64_t events_, int event_version_, int type_);
 ZMQ_EXPORT int zmq_socket_monitor_pipes_stats (void *s);
 
+ZMQ_EXPORT int zmq_ppoll (zmq_pollitem_t *items_, int nitems_, long timeout_, const sigset_t *sigmask_ = nullptr);
```
needs to be duplicated in src/zmq_drafts.cpp

bluca commented:
documentation is missing

bluca commented:
many CI builds are failing

egpbos commented:
Added documentation and added duplicate declaration in zmq_draft.h. Also removed default nullptr sigmask parameter, I think it's more consistent this way, since no other zmq functions seem to have default parameters.

bluca commented on include/zmq.h:
```diff
@@ -763,6 +766,8 @@ ZMQ_EXPORT int zmq_socket_monitor_versioned (
   void *s_, const char *addr_, uint64_t events_, int event_version_, int type_);
 ZMQ_EXPORT int zmq_socket_monitor_pipes_stats (void *s);
 
+ZMQ_EXPORT int zmq_ppoll (zmq_pollitem_t *items_, int nitems_, long timeout_, const sigset_t *sigmask_);
```
It doesn't look like this (sigset_t) is compatible with windows, so I'd suggest to just ifdef both this and the signal.h include

egpbos replied:
Ah right, makes sense, most builds were passing before, so this should fix most of them again!

egpbos replied:
What if, to maintain full API in at least some way, we do this?

```suggestion
#if !defined _WIN32
ZMQ_EXPORT int zmq_ppoll (zmq_pollitem_t *items_, int nitems_, long timeout_, const sigset_t *sigmask_);
#else
ZMQ_EXPORT int zmq_ppoll (zmq_pollitem_t *items_, int nitems_, long timeout_, const void *sigmask_);
#endif
```

This way, it's still callable, but it will just return with errno ENOTSUP anyway on Windows, so the type doesn't really matter.

egpbos replied:
I implemented this in the latest force push.

bluca commented on include/zmq.h:
```diff
@@ -101,6 +101,9 @@ typedef unsigned __int8 uint8_t;
 #endif
 
+// needed for sigset_t definition
+#include <signal.h>
```
This should be ifdefd for windows too just in case

egpbos replied:
Will do.

bluca commented:
Builds are still failing

egpbos commented:
Yes, I'm going to look into failing builds now.

egpbos commented:
- One of the failures seems to be from some C++11 features I used. Converted relevant parts to C++98.
- Another was a missing include in the `test_zmq_ppoll_signals.cpp`.

Remaining [Windows build failures](https://ci.appveyor.com/project/zeromq/libzmq/builds/40863013) I think are unrelated to this PR:
- The first one (on x64) fails `test_monitor`, which I don't think I touched any parts of.
- The cygwin64 build seems to fail on some sort of external server timeout.
- The mingw64 failure fails `test_srcfd` and `test_unbind_wildcard`, which I also didn't touch.

I don't understand the fuzzer builds, do you see whether those are working as expected?

bluca commented:
Yes this is somehow breaking the fuzzers build, and it needs to be fixed

bluca commented:
never mind, I've seen it fail in another PR too now

egpbos commented:
Latest force push fixes yet another C++11 feature that was failing CentOS_6 builds. Other than this, it seems everything is passing now (or explained by something other than this PR).

bluca commented:
Can you rebase on latest master? Half of the CI was down due to Travis

egpbos commented:
Rebased on master.

bluca commented:
There are new build failures to fix

egpbos commented:
I wasn't able to fix everything yet, but at least some failures were due to building the ppoll tests with draft mode disabled, fixed this in the CMake and autotools files.

Two other failures are puzzling, configurations:

-  build (ubuntu-latest, cmake, libsodium, enabled, cmake libsodium-dev, enabled) 
-  build (ubuntu-latest, cmake, clang-tidy clang-tools, tweetnacl, enabled, clang++) 

These builds have drafts enabled, but still complain that `zmq_ppoll` is missing:

```
/home/runner/work/libzmq/libzmq/libzmq/zeromq-4.3.5/tests/test_zmq_ppoll_fd.cpp:72:27: error: use of undeclared identifier 'zmq_ppoll'
```

The common factor that differs from the passing ubuntu builds seems to be cmake. Could it be that cmake doesn't correctly pass on the draft enabling?

I will fix the two Windows regressions, I know what's happening there. Also the clang-format build.

egpbos commented:
@bluca is it possible that something went wrong with the added CI jobs on GitHub Actions? I see in CI.yaml a lot of jobs that have a DRAFT variable, but nothing is done with that variable in the cmake run step. Perhaps this should be ENABLE_DRAFTS? Or am I missing something?

egpbos commented:
Looks like the CI configuration may indeed be an issue. The fix I applied, which should have fixed non-draft builds only, also worked for the failing draft builds (only failing build on GH actions is now the clang-format one).

egpbos commented:
Last push should fix clang-format build.

egpbos commented:
Two remaining failures in three builds:

1. [Debian_9.0/i586](https://build.opensuse.org/package/live_build_log/home:bluca:network:messaging:zeromq:git-draft:PR-4255/libzmq/Debian_9.0/i586) and [Fedora_34/s390x](https://build.opensuse.org/package/live_build_log/home:bluca:network:messaging:zeromq:git-draft:PR-4255/libzmq/Fedora_34/s390x) fail on unrelated `test_ws_transport`
2. [CentOS_7/x86_64](https://build.opensuse.org/package/live_build_log/home:bluca:network:messaging:zeromq:git-draft:PR-4255/libzmq/CentOS_7/x86_64) seems to have hung. It's possible this happens when the 1 second sleep on the child process side isn't enough for the parent side to complete its recv. I will make the test more robust to this failure, since it doesn't matter anyway for the test of ppoll. The ppoll calls have already completed succesfully by this point.

egpbos commented:
Added fix for hanging CentOS_7/x86_64 build, all builds should now pass, except the `test_ws_transport` ones, which I'm pretty sure should not be caused by this PR.

egpbos commented:
The previously failing builds now succeed.

Fuzzers failing now, but this seems spurious and happens on other recent PRs as well. They did pass on the previous run.

Ready to merge?
